File: impl/kotlin/src/test/kotlin/AguiTests.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File
import kotlin.system.exitProcess

//
// C21 NAALP-AGUI UI-consent-binding known-answer test for the Kotlin SDK (design.md §24; R-AGUI-1..6),
// graded against the shared independent corpus vectors/agui/cases.json (NOT produced by this code).
//
// CORPUS-GRADED (pure, no signature): each shown tool-lifecycle event body + head + content id; the
// receipt-chained shown-chain walk (each running head + the final head); an oversized seq (0x0102030405060708
// > 2^53) round-tripping byte-exact (carried as a 64-bit integer, no float rounding); the minimal event;
// the edge cases (non-canonical key order rejected, empty-vs-absent action, a body missing its field-5
// chain back-pointer rejected); a removed shown-event detected as a HOLE with its position; and an unknown
// event kind rejected. ISOLATION (real FIPS-204 ML-DSA-65 via BouncyCastle — Kotlin is a full-signature
// port; the corpus carries NO signed vector, so these are demonstrated with a fixed local seed, stated
// honestly, NOT corpus-graded): the signed shown-chain verifies; and the consent binding — a human §7
// approval binds the EXACT action shown-and-approved, so executing a SUBSTITUTED action (different content
// id) is ActionSubstituted, and a chain with no approved event is UINoConsent — reusing the §7 Approval
// UNCHANGED.
//
// KAT convention: a standalone main() exiting non-zero on any failure; the filename carries the "Tests"
// token the ten-language-parity gate indexes. Written test-first: [Agui] is absent until Agui.kt lands, so
// this fails RED with a kotlinc "unresolved reference: Agui". The load-bearing mutation — the detectHole
// hole return made a no-hole constant (return (i, true) -> return (0, false)) — flips "agui hole detected
// at position 1" on its assertion.
//

private var agFails = 0

private fun agCheck(name: String, got: String, want: String) {
    if (got == want) {
        println("  ok   $name")
    } else {
        agFails++
        println("  FAIL $name\n       got  $got\n       want $want")
    }
}

// ---- balanced-brace / regex JSON access (no JSON library on the Kotlin port) ----

private fun agFindVector(): File {
    var d: File? = File(".").absoluteFile
    repeat(6) {
        val cur = d ?: throw AssertionError("vectors/agui/cases.json not found")
        val p = File(File(cur, "vectors"), "agui/cases.json")
        if (p.isFile) return p
        d = cur.parentFile
    }
    throw AssertionError("vectors/agui/cases.json not found from ${File(".").absolutePath}")
}

private fun agSection(scope: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*").find(scope)
        ?: throw AssertionError("key not found: $key")
    var i = m.range.last + 1
    while (i < scope.length && scope[i].isWhitespace()) i++
    val open = scope[i]
    val close = when (open) { '{' -> '}'; '[' -> ']'; else -> throw AssertionError("not object/array at $key") }
    var depth = 0
    var inStr = false
    var esc = false
    val start = i
    while (i < scope.length) {
        val c = scope[i]
        if (inStr) {
            when {
                esc -> esc = false
                c == '\\' -> esc = true
                c == '"' -> inStr = false
            }
        } else {
            when (c) {
                '"' -> inStr = true
                open -> depth++
                close -> { depth--; if (depth == 0) return scope.substring(start, i + 1) }
            }
        }
        i++
    }
    throw AssertionError("unbalanced value at key: $key")
}

private fun agElements(arraySection: String): List<String> {
    val out = ArrayList<String>()
    var i = 1
    var inStr = false
    var esc = false
    var depth = 0
    var start = -1
    while (i < arraySection.length - 1) {
        val c = arraySection[i]
        if (inStr) {
            when {
                esc -> esc = false
                c == '\\' -> esc = true
                c == '"' -> inStr = false
            }
        } else {
            when (c) {
                '"' -> inStr = true
                '{' -> { if (depth == 0) start = i; depth++ }
                '}' -> { depth--; if (depth == 0 && start >= 0) { out.add(arraySection.substring(start, i + 1)); start = -1 } }
            }
        }
        i++
    }
    return out
}

private fun agStr(scope: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*\"([^\"]*)\"").find(scope)
        ?: throw AssertionError("string key not found: $key")
    return m.groupValues[1]
}

private fun agInt(scope: String, key: String): Long {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*(\\d+)").find(scope)
        ?: throw AssertionError("int key not found: $key")
    return m.groupValues[1].toLong()
}

private fun agHex(s: String): ByteArray = Hex.decode(s)

private const val AG_ALG = Cose.ALG_MLDSA65
private val AG_SEED = ByteArray(32) { 4 }
private val AG_PK = Cose.mldsaKeygen("ML-DSA-65", AG_SEED)
private val AG_APPR_SEED = ByteArray(32) { 8 }
private val AG_APPR_PK = Cose.mldsaKeygen("ML-DSA-65", AG_APPR_SEED)

private inline fun agKind(block: () -> Unit): String = try {
    block(); "no-error"
} catch (e: NaalpException) {
    e.kind
}

private fun aguiRun() {
    val json = agFindVector().readText(Charsets.UTF_8)
    val session = agHex(agStr(json, "session_hex"))

    // 1. the receipt-chained shown chain: each event body/head/id, the running heads, the final head.
    val chainSec = agSection(json, "chain")
    val evElems = agElements(agSection(chainSec, "events"))
    val events = ArrayList<Agui.UIEvent>()
    for (eb in evElems) {
        val e = Agui.UIEvent(session, agInt(eb, "kind"), agHex(agStr(eb, "action_hex")), agInt(eb, "seq"), agHex(agStr(eb, "prev_hex")))
        agCheck("chain event seq=${agInt(eb, "seq")} body == oracle", Hex.encode(e.bytes()), agStr(eb, "body_hex"))
        agCheck("chain event seq=${agInt(eb, "seq")} head == oracle", Hex.encode(e.head()), agStr(eb, "head_hex"))
        agCheck("chain event seq=${agInt(eb, "seq")} id == oracle", Hex.encode(e.id()), agStr(eb, "id_hex"))
        events.add(e)
    }
    val shown = Agui.walkShown(events)
    for (i in evElems.indices) {
        agCheck("walkShown running head [$i] == oracle", Hex.encode(shown[i].head), agStr(evElems[i], "head_hex"))
    }
    agCheck("walkShown final head == oracle", Hex.encode(shown[shown.size - 1].head), agStr(chainSec, "final_head_hex"))

    // 2. an oversized seq (>2^53) round-trips byte-exact — carried as a 64-bit integer, no float rounding.
    val bigSec = agSection(json, "big_seq")
    val bigSeq = agStr(bigSec, "seq_str").toLong()
    val bigEv = Agui.UIEvent(session, agInt(bigSec, "kind"), agHex(agStr(bigSec, "action_hex")), bigSeq, agHex(agStr(bigSec, "prev_hex")))
    agCheck("big_seq body == oracle", Hex.encode(bigEv.bytes()), agStr(bigSec, "body_hex"))
    agCheck("big_seq head == oracle", Hex.encode(bigEv.head()), agStr(bigSec, "head_hex"))
    agCheck("big_seq id == oracle", Hex.encode(bigEv.id()), agStr(bigSec, "id_hex"))

    // 3. the minimal event.
    val minSec = agSection(json, "minimal")
    val minEv = Agui.UIEvent(agHex(agStr(minSec, "session_hex")), agInt(minSec, "kind"), agHex(agStr(minSec, "action_hex")), agInt(minSec, "seq"), agHex(agStr(minSec, "prev_hex")))
    agCheck("minimal body == oracle", Hex.encode(minEv.bytes()), agStr(minSec, "body_hex"))
    agCheck("minimal head == oracle", Hex.encode(minEv.head()), agStr(minSec, "head_hex"))
    agCheck("minimal id == oracle", Hex.encode(minEv.id()), agStr(minSec, "id_hex"))

    // 4. edge cases.
    val edge = agSection(json, "edge_cases")
    val koo = agSection(edge, "keys_out_of_order")
    agCheck("edge keys_out_of_order noncanonical rejected", agKind { Cbor.decode(agHex(agStr(koo, "noncanonical_body_hex"))) }, agStr(koo, "reject"))
    agCheck("edge keys_out_of_order canonical parses", agKind { Agui.parseUIEvent(agHex(agStr(koo, "canonical_body_hex"))) }, "no-error")

    val eva = agSection(edge, "empty_vs_absent")
    val emptyAct = agSection(eva, "empty_action")
    agCheck("edge empty_action id == oracle", Hex.encode(Agui.parseUIEvent(agHex(agStr(emptyAct, "body_hex"))).id()), agStr(emptyAct, "id_hex"))
    val popAct = agSection(eva, "populated_action")
    agCheck("edge populated_action id == oracle", Hex.encode(Agui.parseUIEvent(agHex(agStr(popAct, "body_hex"))).id()), agStr(popAct, "id_hex"))
    val absentF = agSection(eva, "absent_field")
    agCheck("edge absent_field rejected", agKind { Agui.parseUIEvent(agHex(agStr(absentF, "body_hex"))) }, agStr(absentF, "reject"))

    val look = agSection(edge, "look_alike")
    agCheck("edge look_alike (missing prev) rejected", agKind { Agui.parseUIEvent(agHex(agStr(look, "body_hex"))) }, agStr(look, "reject"))

    // 5. a removed shown-event is a detectable HOLE with its position: present indices ev0, ev2 (ev1 omitted)
    //    break contiguity at position 1. THIS is the mutation's target assertion.
    val holeSec = agSection(json, "hole")
    val (holePos, holeFound) = Agui.detectHole(listOf(events[0], events[2]))
    agCheck("agui hole found", holeFound.toString(), "true")
    agCheck("agui hole detected at position 1", holePos.toString(), agInt(holeSec, "position").toString())

    // 6. an unknown event kind is rejected.
    val unknownKind = agInt(json, "unknown_kind")
    val unkEv = Agui.UIEvent(session, unknownKind, agHex(agStr(json, "action_cid_hex")), 0L, Agui.genesis())
    agCheck("agui unknown kind rejected", agKind { Agui.walkShown(listOf(unkEv)) }, "UnknownUIEventKind")

    // 7. ISOLATION (real ML-DSA): the signed shown-chain and the consent binding.
    val now = 1000L
    val actionCid = agHex(agStr(json, "action_cid_hex"))
    val actionBytes = agHex(agStr(json, "action_bytes_hex"))
    val substitutedBytes = agHex(agStr(json, "substituted_bytes_hex"))
    // exported action content id matches the oracle (a relying party computes it over the exact bytes).
    agCheck("iso action content id == oracle", Hex.encode(Agui.contentId(actionBytes)), agStr(json, "action_cid_hex"))

    // 7a. the signed shown-chain verifies under the UI authority's key.
    val signedObjs = events.map { Agui.signUIEvent(it, AG_ALG, AG_SEED) }
    agCheck("iso signed shown-chain verifies", agKind { Agui.verifyShownChain(signedObjs, Cose.PROFILE_PUBLIC, AG_ALG, AG_PK) }, "no-error")
    val tampered = signedObjs[1].copyOf(); tampered[tampered.size - 1] = (tampered[tampered.size - 1].toInt() xor 1).toByte()
    agCheck("iso tampered shown-chain rejected BadSignature", agKind { Agui.verifyShownChain(listOf(signedObjs[0], tampered, signedObjs[2]), Cose.PROFILE_PUBLIC, AG_ALG, AG_PK) }, "BadSignature")

    // 7b. consent binding: a chain that shows + approves actionCid; the human approval binds actionCid.
    val e0 = Agui.UIEvent(session, Agui.KIND_SHOWN, actionCid, 0L, Agui.genesis())
    val e1 = Agui.UIEvent(session, Agui.KIND_ARGS_SHOWN, actionCid, 1L, e0.head())
    val e2 = Agui.UIEvent(session, Agui.KIND_APPROVED, actionCid, 2L, e1.head())
    val consentChain = listOf(e0, e1, e2)
    val appr = Approval.ApprovalRecord(actionCid, "human", Policy.DESTRUCTIVE, ByteArray(16) { 2 }, 2_000_000_000_000L)
    val apprSig = Approval.signApproval(appr, AG_ALG, AG_APPR_SEED)
    agCheck("iso consent binds the exact action shown+approved", agKind { Agui.verifyConsent(consentChain, actionBytes, appr, AG_ALG, AG_APPR_PK, apprSig, now) }, "no-error")
    // a substituted action (different content id) is rejected.
    agCheck("iso substituted action rejected ActionSubstituted", agKind { Agui.verifyConsent(consentChain, substitutedBytes, appr, AG_ALG, AG_APPR_PK, apprSig, now) }, "ActionSubstituted")
    // a chain with no approved event has no consent to bind.
    val noApprove = listOf(e0, e1)
    agCheck("iso no approved event rejected UINoConsent", agKind { Agui.verifyConsent(noApprove, actionBytes, appr, AG_ALG, AG_APPR_PK, apprSig, now) }, "UINoConsent")
}

fun main() {
    println("agui conformance (Kotlin) — graded vs vectors/agui/cases.json")
    aguiRun()
    println(if (agFails == 0) "AguiTests: PASS" else "AguiTests: FAIL ($agFails)")
    exitProcess(if (agFails == 0) 0 else 1)
}

File: impl/csharp/test/AguiTests.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// C21 NAALP-AGUI UI-consent-binding conformance for the C# SDK (design.md §24; R-AGUI-1..6),
    /// graded against the shared independent corpus <c>vectors/agui/cases.json</c> (NOT produced by this
    /// code): the receipt-chained shown-event bodies/heads/ids, the running and final chain heads, the
    /// genesis prev, the >2^53 seq carried byte-exact (big_seq), the minimal and empty-vs-absent bodies,
    /// the non-canonical / malformed rejections, and the omitted-shown-event HOLE detected at its
    /// POSITION.
    ///
    /// <para>The signed shown-chain and the UI consent binding — which REUSES the just-landed C#
    /// <see cref="Approval"/> §7 approval UNCHANGED — use REAL deterministic ML-DSA-65 (BouncyCastle,
    /// rnd=0) and are demonstrated in isolation on the corpus action bytes (the corpus carries no signed
    /// vector; F2/F4, honest): a valid consent binds the shown-and-approved action; a SUBSTITUTED action
    /// (different content id) is rejected ActionSubstituted; a chain with no approved event is
    /// UINoConsent. Ported from impl/go/agui; mirrors impl/python/naalp/agui.py.</para>
    /// </summary>
    public sealed class AguiTests
    {
        private const int Alg = Cose.ALG_MLDSA65;

        private static JsonElement Vector()
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            for (int i = 0; i < 12 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, "vectors", "agui", "cases.json");
                if (File.Exists(p))
                {
                    return JsonDocument.Parse(File.ReadAllText(p)).RootElement.Clone();
                }
                d = d.Parent;
            }
            throw new Xunit.Sdk.XunitException("vectors/agui/cases.json not found from " + AppContext.BaseDirectory);
        }

        private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();
        private static byte[] Hb(string s) => Convert.FromHexString(s);

        private static Agui.UIEvent EventOf(JsonElement ev, byte[] session)
        {
            return new Agui.UIEvent(
                session,
                ev.GetProperty("kind").GetInt64(),
                Hb(ev.GetProperty("action_hex").GetString()!),
                ev.GetProperty("seq").GetInt64(),
                Hb(ev.GetProperty("prev_hex").GetString()!));
        }

        // ---- the receipt-chained shown-event bodies, running heads, and final head -----------------
        // Also grades the closed genesis (48 zero bytes) and each event's content id.

        [Fact]
        public void ChainBodiesAndHeadsMatchOracle()
        {
            JsonElement v = Vector();
            byte[] session = Hb(v.GetProperty("session_hex").GetString()!);
            Assert.Equal(v.GetProperty("genesis_hex").GetString(), Hex(Agui.Genesis()));

            JsonElement chain = v.GetProperty("chain");
            byte[] running = Agui.Genesis();
            int count = 0;
            foreach (JsonElement ev in chain.GetProperty("events").EnumerateArray())
            {
                Agui.UIEvent e = EventOf(ev, session);
                Assert.Equal(ev.GetProperty("prev_hex").GetString(), Hex(running)); // prev links to the running head
                Assert.Equal(ev.GetProperty("body_hex").GetString(), Hex(e.Bytes()));
                Assert.Equal(ev.GetProperty("head_hex").GetString(), Hex(e.Head()));
                Assert.Equal(ev.GetProperty("id_hex").GetString(), Hex(e.Id()));
                running = e.Head();
                count++;
            }
            Assert.Equal(chain.GetProperty("final_head_hex").GetString(), Hex(running));
            Assert.True(count > 0, "corpus carried no chain events");
        }

        // ---- a >2^53 seq round-trips byte-exact (carried as a decimal string, not a JSON number) ---

        [Fact]
        public void BigSeqByteExact()
        {
            JsonElement v = Vector();
            byte[] session = Hb(v.GetProperty("session_hex").GetString()!);
            JsonElement bs = v.GetProperty("big_seq");
            long seq = long.Parse(bs.GetProperty("seq_str").GetString()!); // 72623859790382856 > 2^53
            Assert.True(seq > (1L << 53));
            var e = new Agui.UIEvent(session, bs.GetProperty("kind").GetInt64(), Hb(bs.GetProperty("action_hex").GetString()!),
                seq, Hb(bs.GetProperty("prev_hex").GetString()!));
            Assert.Equal(bs.GetProperty("body_hex").GetString(), Hex(e.Bytes()));
            Assert.Equal(bs.GetProperty("head_hex").GetString(), Hex(e.Head()));
            Assert.Equal(bs.GetProperty("id_hex").GetString(), Hex(e.Id()));
        }

        // ---- minimal, empty-vs-absent, non-canonical, and look-alike edge cases --------------------

        [Fact]
        public void EdgeCasesMatchOracle()
        {
            JsonElement v = Vector();

            JsonElement min = v.GetProperty("minimal");
            var minE = new Agui.UIEvent(Hb(min.GetProperty("session_hex").GetString()!), min.GetProperty("kind").GetInt64(),
                Hb(min.GetProperty("action_hex").GetString()!), min.GetProperty("seq").GetInt64(), Hb(min.GetProperty("prev_hex").GetString()!));
            Assert.Equal(min.GetProperty("body_hex").GetString(), Hex(minE.Bytes()));
            Assert.Equal(min.GetProperty("head_hex").GetString(), Hex(minE.Head()));
            Assert.Equal(min.GetProperty("id_hex").GetString(), Hex(minE.Id()));

            JsonElement eva = v.GetProperty("edge_cases").GetProperty("empty_vs_absent");
            byte[] session = Hb(v.GetProperty("session_hex").GetString()!);

            // an empty action bstr is PRESENT and valid, distinct by content id from a populated one.
            var emptyAction = new Agui.UIEvent(session, 0, Array.Empty<byte>(), 0, Agui.Genesis());
            Assert.Equal(eva.GetProperty("empty_action").GetProperty("id_hex").GetString(), Hex(emptyAction.Id()));
            var populated = new Agui.UIEvent(session, 0, Hb(eva.GetProperty("populated_action").GetProperty("action_hex").GetString()!), 0, Agui.Genesis());
            Assert.Equal(eva.GetProperty("populated_action").GetProperty("id_hex").GetString(), Hex(populated.Id()));
            Assert.NotEqual(Hex(emptyAction.Id()), Hex(populated.Id()));

            // a body whose mandatory field 3 (action) is ABSENT is rejected UIMalformed.
            var abEx = Assert.Throws<NaalpException>(() => Agui.ParseUIEvent(Hb(eva.GetProperty("absent_field").GetProperty("body_hex").GetString()!)));
            Assert.Equal(eva.GetProperty("absent_field").GetProperty("reject").GetString(), abEx.Kind);

            // a ui-event-shaped body lacking its field-5 back-pointer (prev) is rejected UIMalformed.
            JsonElement la = v.GetProperty("edge_cases").GetProperty("look_alike");
            var laEx = Assert.Throws<NaalpException>(() => Agui.ParseUIEvent(Hb(la.GetProperty("body_hex").GetString()!)));
            Assert.Equal(la.GetProperty("reject").GetString(), laEx.Kind);

            // keys emitted descending -> the strict decoder rejects (surfaced as UIMalformed by the parser).
            JsonElement koo = v.GetProperty("edge_cases").GetProperty("keys_out_of_order");
            var ncEx = Assert.Throws<NaalpException>(() => Agui.ParseUIEvent(Hb(koo.GetProperty("noncanonical_body_hex").GetString()!)));
            Assert.Equal("UIMalformed", ncEx.Kind);
            // the canonical form parses cleanly.
            Agui.ParseUIEvent(Hb(koo.GetProperty("canonical_body_hex").GetString()!));
        }

        // ---- an unknown kind is rejected; a walked chain enforces contiguity -----------------------

        [Fact]
        public void KindVocabularyAndWalk()
        {
            JsonElement v = Vector();
            byte[] session = Hb(v.GetProperty("session_hex").GetString()!);
            foreach (JsonElement k in v.GetProperty("kind_vocabulary").EnumerateArray())
            {
                Assert.True(Agui.IsKnownKind(k.GetProperty("code").GetInt64()));
            }
            Assert.False(Agui.IsKnownKind(v.GetProperty("unknown_kind").GetInt64())); // 99

            // walking the honest 3-event chain succeeds and preserves the shown action ids.
            var full = new List<Agui.UIEvent>();
            foreach (JsonElement ev in v.GetProperty("chain").GetProperty("events").EnumerateArray())
            {
                full.Add(EventOf(ev, session));
            }
            List<Agui.ShownEvent> shown = Agui.WalkShown(full);
            Assert.Equal(full.Count, shown.Count);
            Assert.Equal(v.GetProperty("chain").GetProperty("final_head_hex").GetString(), Hex(shown[shown.Count - 1].Head));
        }

        // ---- an omitted shown-event is a detectable HOLE at its POSITION ---------------------------
        // MUTATION ANCHOR: DetectHole returning (0,false) fails Assert.True(hole) — the gap-evidence
        // property (a removed shown-event leaves a hole where it should have been).

        [Fact]
        public void HoleDetectedAtPosition()
        {
            JsonElement v = Vector();
            byte[] session = Hb(v.GetProperty("session_hex").GetString()!);
            JsonElement events = v.GetProperty("chain").GetProperty("events");

            // the gappy chain: present ev0 and ev2, the middle shown-event (ev1) omitted.
            var gappy = new List<Agui.UIEvent> { EventOf(events[0], session), EventOf(events[2], session) };

            (int pos, bool hole) = Agui.DetectHole(gappy);
            Assert.True(hole);
            Assert.Equal(v.GetProperty("hole").GetProperty("position").GetInt32(), pos);

            // the contiguous full chain has no hole.
            var full = new List<Agui.UIEvent> { EventOf(events[0], session), EventOf(events[1], session), EventOf(events[2], session) };
            (_, bool fullHole) = Agui.DetectHole(full);
            Assert.False(fullHole);

            // WalkShown rejects the gappy chain (fail-closed) — the same broken link.
            var wex = Assert.Throws<NaalpException>(() => Agui.WalkShown(gappy));
            Assert.Equal("UIChainBroken", wex.Kind);
        }

        // ---- the signed shown-chain + the UI consent binding (REAL ML-DSA, isolation) --------------

        [Fact]
        public void ConsentBindsExactShownAction()
        {
            JsonElement v = Vector();
            byte[] session = Hb(v.GetProperty("session_hex").GetString()!);
            byte[] actionCid = Hb(v.GetProperty("action_cid_hex").GetString()!);
            byte[] actionBytes = Hb(v.GetProperty("action_bytes_hex").GetString()!);
            byte[] substituted = Hb(v.GetProperty("substituted_bytes_hex").GetString()!);

            // non-circular: the action content ids come from the corpus, recomputed here from the bytes.
            Assert.Equal(v.GetProperty("action_cid_hex").GetString(), Hex(Agui.ContentId(actionBytes)));
            Assert.Equal(v.GetProperty("substituted_cid_hex").GetString(), Hex(Agui.ContentId(substituted)));

            // the UI authority signs the shown chain (shown, args-shown, approved).
            byte[] uiSeed = new byte[32];
            for (int i = 0; i < 32; i++) uiSeed[i] = 41;
            byte[] uiPk = Cose.MldsaKeygen("ML-DSA-65", uiSeed);
            var objs = new List<byte[]>();
            var chain = new List<Agui.UIEvent>();
            foreach (JsonElement ev in v.GetProperty("chain").GetProperty("events").EnumerateArray())
            {
                Agui.UIEvent e = EventOf(ev, session);
                chain.Add(e);
                objs.Add(Agui.SignUIEvent(e, Alg, uiSeed));
            }
            List<Agui.UIEvent> verified = Agui.VerifyShownChain(objs, Cose.PROFILE_PUBLIC, Alg, uiPk);
            Assert.Equal(chain.Count, verified.Count);

            // the human approval binds the shown-and-approved action content id.
            byte[] humanSeed = new byte[32];
            for (int i = 0; i < 32; i++) humanSeed[i] = 43;
            byte[] humanPk = Cose.MldsaKeygen("ML-DSA-65", humanSeed);
            var appr = new Approval.ApprovalRecord(actionCid, "human", Policy.NON_IDEMPOTENT_WRITE, new byte[] { 5, 5 }, 1000);
            byte[] apprSig = Approval.SignApproval(appr, Alg, humanSeed);

            // honest consent: the executed action IS the shown-and-approved one -> authorized.
            Agui.VerifyConsent(chain, actionBytes, appr, Alg, humanPk, apprSig, 500);

            // a SUBSTITUTED action (different content id) is rejected.
            var subEx = Assert.Throws<NaalpException>(() => Agui.VerifyConsent(chain, substituted, appr, Alg, humanPk, apprSig, 500));
            Assert.Equal("ActionSubstituted", subEx.Kind);

            // a chain with no approved event has no consent to bind.
            var shownOnly = new List<Agui.UIEvent> { chain[0] };
            var ncEx = Assert.Throws<NaalpException>(() => Agui.VerifyConsent(shownOnly, actionBytes, appr, Alg, humanPk, apprSig, 500));
            Assert.Equal("UINoConsent", ncEx.Kind);

            // an expired approval is rejected (the §7 approval expiry, reused unchanged).
            var expEx = Assert.Throws<NaalpException>(() => Agui.VerifyConsent(chain, actionBytes, appr, Alg, humanPk, apprSig, 1001));
            Assert.Equal("ApprovalExpired", expEx.Kind);

            // an approval verified under the WRONG key is BadSignature (fail-closed).
            byte[] foreignSeed = new byte[32];
            for (int i = 0; i < 32; i++) foreignSeed[i] = 88;
            byte[] foreignPk = Cose.MldsaKeygen("ML-DSA-65", foreignSeed);
            var bsEx = Assert.Throws<NaalpException>(() => Agui.VerifyConsent(chain, actionBytes, appr, Alg, foreignPk, apprSig, 500));
            Assert.Equal("BadSignature", bsEx.Kind);
        }
    }
}
